core/lib/src/fs/server.rs

```diff
@@ -60,351 +72,492 @@ use crate::fs::NamedFile;
-                        .into_owned();
-
-                    return Redirect::permanent(normal)
-                        .respond_to(req)
-                        .or_forward((data, Status::InternalServerError));
-                }
-
-                if !options.contains(Options::Index) {
-                    return Outcome::forward(data, Status::NotFound);
-                }
-
-                let index = NamedFile::open(p.join("index.html")).await;
-                index.respond_to(req).or_forward((data, Status::NotFound))
-            },
-            Some(p) => {
-                let file = NamedFile::open(p).await;
-                file.respond_to(req).or_forward((data, Status::NotFound))
-            }
-            None => Outcome::forward(data, Status::NotFound),
+            Some(FileResponse::File(file)) => Some(self.0(file)),
+            path => path,
         }
     }
 }
 
-/// A bitset representing configurable options for [`FileServer`].
+/// Prepends the provided path, to serve files from a directory.
+///
+/// You can use [`relative!`] to make a path relative to the crate root, rather
+/// than the runtime directory.
 ///
-/// The valid options are:
+/// # Example
+///
+/// ```rust,no_run
+/// # use rocket::fs::{FileServer, dir_root, relative};
+/// # fn make_server() -> FileServer {
+/// FileServer::empty()
+///     .map_file(dir_root(relative!("static")))
+/// # }
+/// ```
 ///
-///   * [`Options::None`] - Return only present, visible files.
-///   * [`Options::DotFiles`] - In addition to visible files, return dotfiles.
-///   * [`Options::Index`] - Render `index.html` pages for directory requests.
-///   * [`Options::IndexFile`] - Allow serving a single file as the index.
-///   * [`Options::Missing`] - Don't fail if the path to serve is missing.
-///   * [`Options::NormalizeDirs`] - Redirect directories without a trailing
-///     slash to ones with a trailing slash.
+/// # Panics
 ///
-/// `Options` structures can be `or`d together to select two or more options.
-/// For instance, to request that both dot files and index pages be returned,
-/// use `Options::DotFiles | Options::Index`.
-#[derive(Debug, Clone, Copy)]
-pub struct Options(u8);
+/// Panics if `path` is not directory.
+pub fn dir_root(path: impl AsRef<Path>)
+    -> impl for<'p, 'h> Fn(File<'p, 'h>) -> FileResponse<'p, 'h> + Send + Sync + 'static
+{
+    use yansi::Paint as _;
 
-#[allow(non_upper_case_globals, non_snake_case)]
-impl Options {
-    /// All options disabled.
-    ///
-    /// Note that this is different than [`Options::default()`](#impl-Default),
-    /// which enables options.
-    pub const None: Options = Options(0);
+    let path = path.as_ref();
+    if !path.is_dir() {
+        let path = path.display();
+        error!("FileServer path '{}' is not a directory.", path.primary());
+        warn!("Aborting early to prevent inevitable handler error.");
+        panic!("invalid directory: refusing to continue");
```

Comment:
This will need to be rewritten to be structured `trace` messages.

Reply:
I've updated it (and removed yansi in the process). If there is more information we can provide, I'd be more than happy to update these messages.